pinot-core/src/main/java/com/linkedin/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java

```diff
@@ -355,18 +352,17 @@ protected boolean consumeLoop() throws Exception {
     return true;
   }
 
-  private void processKafkaEvents(Iterable<MessageAndOffset> messagesAndOffsets, Long highWatermark, long idlePipeSleepTimeMillis) {
+  private void processKafkaEvents(MessageBatch messagesAndOffsets, long idlePipeSleepTimeMillis) {
     Meter realtimeRowsConsumedMeter = null;
     Meter realtimeRowsDroppedMeter = null;
 
-    Iterator<MessageAndOffset> msgIterator = messagesAndOffsets.iterator();
-
     int indexedMessageCount = 0;
     int kafkaMessageCount = 0;
     boolean canTakeMore = true;
     GenericRow decodedRow = null;
     GenericRow transformedRow = null;
-    while (!_shouldStop && !endCriteriaReached() && msgIterator.hasNext()) {
+    int index = 0;
+    while (!_shouldStop && !endCriteriaReached() && index < messagesAndOffsets.getMessageCount()) {
```

Comment:
I would change this to a 'for' loop to protect against adding a 'continue' statement somewhere in the body of the loop.

Reply:
done